### ISSUE TITLE
Dot net core version

### ISSUE THREAD

**Issue by SHAliakbari** [Uriil/MetaTrader4.Manager.Wrapper#113]
Hi, 

With DotnetCore we can build apps cross platform , but i used your library and stuck in windows only . 
Is it possible to create wrapper on mtmanapi to work os independent ?

**Comment by Uriil** [comment]
Microsoft does not have plans to support C++/CLI in .NET core, so there is no way to port current project to .NET core. Maybe there is some workaround, but I don't have any idea about it

**Comment by SHAliakbari** [comment]
Dot net core support p/invoke . Is it possible to use mtmanapi with p/invoke ?
If possible , we can create a dot net core class library without using c++ .


Sent from my Samsung device


-------- Original message --------

**Comment by Uriil** [comment]
Well, we will have to have some intermediate library to keep reference to manager instance, then we will have to map managed types to unmanaged using PInvoke, it's huge amount of work and personally I have no idea how to do it.

However if you decide to start project and do initial research how to do it I will gladly help you to cover entire API

**Comment by SHAliakbari** [comment]
I think we can do it without intermediate library .

I done this before for another pure c++  application . Just declared my managed types in my application and call c++ methods directly in my c# . I agree with you that huge amount of work should be done , and some R&D should be done , Marshaling issues and .... ;

but the main question remains .

Is mtmanapi works on linux or other platforms or can only run on windows ?!

if yes , we can plan to migrate to  dot net core and platform independent.

________________________________

**Comment by Uriil** [comment]
Well, I am not related to MetaQuotes. So I have no idea if mtmanapi is platform independent or not. And we will have to start our research from this point

**Comment by magick93** [comment]
The MT4 Manager can run on Linux if you use wine, so yes, it should be possible to run this on linux using dotnetcore, and probably some windows dlls, or with wine.

**Comment by kakobrekla** [comment]
Wine should work, but mono would probably be a better solution.

**Comment by magick93** [comment]
Wine and mono are not really comparable. Mono is the opensource alternative .net, whereas Wine is like a windows emulation layer. Several core windows dlls are reproduced by wine. 

Mono is more comparable to DotNetCore. So, to run MetaTrader4.Manager.Wrapper on Linux we will likely need either DotNetCore or Mono, and Wine.

**Comment by SHAliakbari** [comment]
If mtmanapi which own by MetaQuotes use "Visual  c++ runtime " , i think can not run in nature of linux , but if it use pure c++ , using P/Invoke and dotnetcore app can run on every platform .

As i know , meta trader manager api uses visual c++ redistributable . I don't know visual c++ redistributable can install on linux or not !


________________________________

**Comment by magick93** [comment]
Yes, visual c++ redistributable can run very well on linux, https://appdb.winehq.org/objectManager.php?sClass=application&iId=5766

**Comment by miho99** [comment]
I can speak with MQ so if you have any questions I can ask.

**Comment by SHAliakbari** [comment]
Well , I think two things are important :


1- requirements to use manapi on other platform than windows

2- any workaround to use manapi with P/Invoke


but let others ask their question if have

________________________________

**Comment by magick93** [comment]
> any workaround to use manapi with P/Invoke

Unless I am mistaken, PInvoke is just another way to call unmanaged dlls, which is already been done by this project. The point being, I dont think there is anything gained from refactoring to use PInvoke. 

But, please correct me if I am mistaken.

**Comment by SHAliakbari** [comment]
Yes you right , but with current implementation we force to use dot net framework 4.* ; and this is not platform independent .

I plan to use manapi in dot net core on linux .

So i should getting rid of dot net framework 4.* and use dot net core and use P/Invoke to call manapi methods .

Do you have any suggestion to use manapi or current library in dot net core ?

________________________________

**Comment by Uriil** [comment]
I have a start point for you.

There are few MQ examples using mtmanapi.dll. Try to run ManagerAPITrade on linux and see if it works or not. If no, there is no way we can do anything with it. If yes, we can continue discussion.

**Comment by SHAliakbari** [comment]
Most of samples use win32 windows , which can not run on linux i think .

but i check them again .

thanks

________________________________

**Comment by magick93** [comment]
Does MT4 Manager use manapi? I would assume yes.

If yes, then it can work on Linux - but it does require Wine.

**Comment by magick93** [comment]
FYI https://blogs.msdn.microsoft.com/vcblog/2016/03/30/visual-c-for-linux-development/

**Comment by liornabat** [comment]
Hi,
will be happy to participate and help to port to .net core. let me know if someone wants to lead it.

**Comment by m-asgari** [comment]
Hello liornabat , I can Help you please send your message to m.asgari@gmail.com

**Comment by magick93** [comment]
We have a lot of people interested in this feature. I think it would be
good if we kept the communication open and on this forum, if possible.

**Comment by SHAliakbari** [comment]
very well , as it shows , we can start working on this . 
I start some research and a test project to determine deep of work , at start i face a problem . 
I think the main work is marshaling . objects are on c lib which should marshal to managed code . in current version this works done by c++ managed wrapper , but when we want to use P/Invoke , marshaling come in front you with anger face !! 

We need a guy with high acknowledge of marshaling between c and managed code , with experience of handling callbacks , object references and ... 

Is anybody there ?!

**Comment by liornabat** [comment]
there is a tool called swig that I played with it yesterday. swig can take your dll and build a wrapper in many languages. http://www.swig.org/

**Comment by magick93** [comment]
While converting to use dotnetcore would mean `MetaTrader4.Manager.Wrapper` is more likely to run on docker, does anyone know if there are already windows docker images which would work? Specifically, are there any docker containers that allow running c++ unmanaged dlls, such as `manapi.dll`?

**Comment by liornabat** [comment]
Any windows container that support full .Net Framework can run the wrapper
The only issue is the size of the image - ~10gig ….. because the full .net require full servercore. Nano server image cannot run .net framework only .net core.
If you thinking about mono under docker, it won’t work either, since he cannot load windows dll.
Also, if you think about using wine, think again,
You will need linux container + wine + .net framework – I was unable to get it installed. Not working. So we tow options:

  1.  .Net – only windows containers – big images
  2.  C++ - will work with linux and wine


Lior,

**Comment by magick93** [comment]
> The only issue is the size of the image - ~10gig

Holly cow! Thats insane! Even a windows VM is smaller than that!


> If you thinking about mono under docker, it won’t work either, since he cannot load windows dll.

What about http://www.mono-project.com/docs/advanced/pinvoke/

**Comment by liornabat** [comment]
Ahh… good luck … not easy….



Lior,

**Comment by filipdulic** [comment]
Based on some research I did, the mtmanapi.dll is dependant on WINAPI so Wine is a must.

**Comment by miho99** [comment]
can this be done?
